python/typedb/common/datetime.py

```diff
@@ -0,0 +1,198 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+from datetime import date, datetime, tzinfo
+from typing import Optional
+from zoneinfo import ZoneInfo
+
+NANOS_DIGITS = 9
+NANOS_IN_SECOND = 10 ** NANOS_DIGITS
+MICROS_IN_NANO = 1000
+
+
+class Datetime:
+    """
+    An extension class for ``datetime.datetime`` class to store additional information about nanoseconds.
+    It is split to a timestamp (time zoned or not) based on the number of full seconds and a nanoseconds part.
+    """
+
+    """
+    Initialise a new ``Datetime``.
+
+    :param timestamp_seconds: Amount of full seconds since the epoch.
+    :param subsec_nanos: A number of nanoseconds since the last seconds boundary. Should be between 0 and 999,999,999.
+    :param tz_name: A timezone name. Accepts any format suitable for ``ZoneInfo``, e.g. IANA.
+    :raises ValueError: If subsec_nanos is not within the valid range.
+    :raises ZoneInfoNotFoundError: If the tz_name is invalid.
+    """
+
+    def __init__(self, timestamp_seconds: int, subsec_nanos: int, tz_name: Optional[str] = None,
+                 is_tz_aware_timestamp: bool = False):
+        if not (0 <= subsec_nanos < NANOS_IN_SECOND):
+            raise ValueError("subsec_nanos must be between 0 and 999,999,999")
+
+        self._tz_name = tz_name
+        if self._tz_name is None:
+            self._datetime_of_seconds = datetime.utcfromtimestamp(timestamp_seconds)
+        else:
+            if is_tz_aware_timestamp:
+                self._datetime_of_seconds = datetime.utcfromtimestamp(timestamp_seconds).replace(
+                    tzinfo=ZoneInfo(self._tz_name))
+            else:
+                self._datetime_of_seconds = datetime.fromtimestamp(timestamp_seconds, ZoneInfo(self._tz_name))
+        self._nanos = subsec_nanos
+
+    @property
+    def datetime_of_seconds(self) -> datetime:
```

Comment:
The docs for the class tell:
```
    An extension class for ``datetime.datetime`` class to store additional information about nanoseconds.
    It is split to a timestamp (time zoned or not) based on the number of full seconds and a nanoseconds part.
```
The docs for the method tell:
`Return the original datetime.datetime part, containing data up to seconds`

I am not fully satisfied with `datetime_of_seconds`, but I want to be explicit here. Otherwise, they may think that this `native_datetime` can contain millis! Do you have more ideas?

Reply:
Maybe we call it `datetime_without_nanos`? Because that's what is different - dropping Nanos?

Reply:
Done. 
Typeql types:
```
            "birth-time": "1999-02-26T12:15:05.001234567",
            "current-time": "2024-09-20T16:40:05.001234567 Europe/Belfast",
```
`print(value.datetime_without_nanos)`:
```
1999-02-26 12:15:05.001234
2024-09-20 16:40:05.001234+01:00
```